### PR TITLE
Introduce a lockless critical section.

### PR DESCRIPTION
**Pull request by vdbergh**
When the browser comes out of sleep then there is a race between the tabs which may lead to duplicate notifications. For some reason this only seems to happen on mobile devices.

This issue could be trivially fixed by not displaying duplicate notifications but this seems like a hack.

Here we implement a lockless critical section following an algorithm by Michael Fischer.

It seems impossible to test this without just trying....

### CODE REVIEW

**Comment by noobpwnftw** [comment]
Can it be solved by having unique notification IDs so that the OS/browser notifier does the duplication?

**Comment by vdbergh** [comment]
Weblocks is marked as experimental technology only supported in the latest browsers. So while being very useful it is perhaps a bit early to use them. I think for the application here a simple lockless scheme is sufficient.

**Comment by dav1312** [comment]
> Weblocks is marked as experimental technology only supported in the latest browsers.

I think that "experimental" notice is outdated, we probably use features with less support than this.
Supported in Chrome since 2018 and Firefox since 2021

https://developer.mozilla.org/en-US/docs/Web/API/Web_Locks_API#browser_compatibility
https://caniuse.com/mdn-api_lock

**Comment by vdbergh** [comment]
> > Weblocks is marked as experimental technology only supported in the latest browsers.
> 
> I think that "experimental" notice is outdated, we probably use features with less support than this. Supported in Chrome since 2018 and Firefox since 2021
> 
> https://developer.mozilla.org/en-US/docs/Web/API/Web_Locks_API#browser_compatibility https://caniuse.com/mdn-api_lock
> 
> The only problem that I see is probably Safari that only supports it since last year and we have some users with old versions

Safari would be an issue. There are plenty of people with old iphones.

**Comment by vdbergh** [comment]
I added some comments explaining how the algorithm works.

File: server/fishtest/static/js/application.js [hunk 1]
```diff
@@ -253,3 +253,43 @@ function broadcast(cmd, arg) {
...
+  // rigorously bound the delay necessary to execute (1)(2)(3) because
+  // there is no bound on memory contention (here corresponding to the
+  // setItem instructions). However for us this is not an issue -
+  // 30ms should be sufficient for plenty of tabs.
+  // Another issue is that taking the lock always takes at least 30ms
+  // (with this implementation) even if there is no contention. This is also
+  // not a problem since our application is not time critical.
+  while (true) {
+    let X;
+    // We wait for the lock to clear. Afer 1.5s we assume that the current
```
Comment (dav1312) [inline server/fishtest/static/js/application.js:273]: After*

Comment (vdbergh) [reply server/fishtest/static/js/application.js]: Fixed! Thanks.

**Comment by vdbergh** [comment]
This is an example with triplicate notifications:

IMG-0238

**Comment by vdbergh** [comment]
I think I can test this PR by using an instrumented version fishtest which periodically answers the `/api/actions` api-call
with some fake finished runs.

**Comment by vdbergh** [comment]
I tried this and for now the verdict seems to be that it does not work at all, even extending the 30ms delay to 750ms (tested on safari an ipad). I don't understand this but it seems impossible to debug this. 

For now the best solution seems to be to simply suppress duplicate notifications. Someday I might do a test with weblocks. If it still does not work then there would be an issue with the code.

**Comment by vdbergh** [comment]
https://www.javaer101.com/en/article/35579652.html

It seems that tabs do not see local storage updates at the same time (like memory behind a cache). This invalidates the algorithm. 

It appears on old browsers one can implement real locks via indexeddb (which has transaction support).  https://www.npmjs.com/package/idb-mutex .

**Comment by vdbergh** [comment]
The only sensible solution seems to be to use weblocks when they are supported and ignore the issue otherwise (it only happens in rather extreme cases with many tabs open).

**Comment by vdbergh** [comment]
So it seems weblocks also don’t solve the issue on Safari on iPad (weblocks are supported). I don’t understand what’s wrong.

**Comment by vdbergh** [comment]
Discovered a stupid bug. async_sleep is an async function so one needs to write await async_sleep.

**Comment by vdbergh** [comment]
But it still doesn't work. I conjecture that changes in local storage take some time to propagate to other tabs, especially if the OS is on the borderline of sleeping (this is on an ipad). So in addition to locks one would need memory barriers for local storage.

**Comment by vdbergh** [comment]
I guess the correct way to handle this is via a service worker.

**Comment by vdbergh** [comment]
But it seems they became only very recently supported by Chrome on Android whereas Safari on iOS had them for a long time:  https://caniuse.com/serviceworkers Strange. Usually Chrome is ahead of the pack.

**Pull request closed without merging**